File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        profit = [0]*n 
        dp = [0]*n 
        dp[0] = prices[0]
        for i in range(1,len(prices)):
            price = prices[i]
            minbuy = min(price, dp[i-1])
            dp[i] = minbuy
            profit[i] = price - minbuy
        return max(profit)
        

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int buy = prices[0];
        int max_profit = 0;
        for(int i=1;i<prices.size();i++)
        {
            buy = min(buy, prices[i]);
            max_profit = max(max_profit, prices[i]-buy);
        }
        return max_profit;
    }
};